core/src/main/scala/org/apache/spark/deploy/Client.scala

```diff
@@ -124,38 +127,57 @@ private class ClientEndpoint(
     }
   }
 
-  /* Find out driver status then exit the JVM */
+  /**
+   * Find out driver status then exit the JVM. If the waitAppCompletion is set to true, monitors
+   * the application until it finishes, fails or is killed.
+   */
   def pollAndReportStatus(driverId: String): Unit = {
     // Since ClientEndpoint is the only RpcEndpoint in the process, blocking the event loop thread
     // is fine.
     logInfo("... waiting before polling master for driver state")
     Thread.sleep(5000)
     logInfo("... polling master for driver state")
-    val statusResponse =
-      activeMasterEndpoint.askSync[DriverStatusResponse](RequestDriverStatus(driverId))
-    if (statusResponse.found) {
-      logInfo(s"State of $driverId is ${statusResponse.state.get}")
-      // Worker node, if present
-      (statusResponse.workerId, statusResponse.workerHostPort, statusResponse.state) match {
-        case (Some(id), Some(hostPort), Some(DriverState.RUNNING)) =>
-          logInfo(s"Driver running on $hostPort ($id)")
-        case _ =>
-      }
-      // Exception, if present
-      statusResponse.exception match {
-        case Some(e) =>
-          logError(s"Exception from cluster was: $e")
-          e.printStackTrace()
-          System.exit(-1)
-        case _ =>
-          System.exit(0)
+    while (true) {
```

Comment:
This could block `ClientEndpoint` when `waitAppCompletion=true`?

Reply:
Hey, please pay attention to my comment here. I believe the current implementation could block `ClientEndpoint` because it's a `ThreadSafeRpcEndpoint`. When enabling `waitAppCompletion`, `ClientEndpoint` would actually keep handling message `SubmitDriverResponse` until the application finished. So, `ClientEndpoint` is unable to handle other messages, e.g. `RemoteProcessDisconnected`, `RemoteProcessConnectionError`, at the same time, which breaks the current behaviour. Furthermore, it could also block messages from backup masters, though not fatal in this case.

Reply:
@Ngone51 Apologies, somehow missed this comment. How can I quickly verify this? Could you kindly suggest if you have any pointers on how this can be fixed?

Reply:
We can periodically send a message (e.g. we can send it after ` Thread.sleep(REPORT_DRIVER_STATUS_INTERVAL)
` ) to  `ClientEndpoint` itself to check driver's status.

Reply:
A possible way to verify this is to launch a long running application and then shutdown Master at the middle and see whether `onDisconnected` is called.

Reply:
@Ngone51 I launched a long-running application with flag enabled and disabled. In both cases, I see the following in driver logs:
20/05/09 13:42:59 WARN StandaloneAppClient$ClientEndpoint: Connection to Akshats-MacBook-Pro.local:7077 failed; waiting for master to reconnect...
20/05/09 13:42:59 WARN StandaloneSchedulerBackend: Disconnected from Spark cluster! Waiting for reconnection...

onDisconnected method from StandaloneAppClient.scala is getting called:
![image](https://user-images.githubusercontent.com/31816865/81468361-bb658480-91fc-11ea-87d5-3d00cbf7619f.png)

Reply:
> We can periodically send a message (e.g. we can send it after `Thread.sleep(REPORT_DRIVER_STATUS_INTERVAL)` ) to `ClientEndpoint` itself to check driver's status.

@Ngone51 Thanks for this suggestion. Just to confirm, are you suggesting to do this in lin # 180 in pollAndReportStatus method? Or should we handle this outside?
![image](https://user-images.githubusercontent.com/31816865/81476588-4449e380-9230-11ea-9704-713b84daf2d4.png)

Reply:
> In both cases, I see the following in driver logs. I couldn't find any difference in logs.

Hi @akshatb1 , logs are from  `StandaloneAppClient$ClientEndpoint` and `StandaloneSchedulerBackend` rather than `org.apache.spark.deploy.ClientEndpoint`. Can you check again?

> Just to confirm, are you suggesting to do this in lin # 180 in pollAndReportStatus method? Or should we handle this outside?

I think just after line #180 should be ok.

Reply:
@Ngone51 Yes, not sure about the logs. I will check again. This is the command I am using to submit jobs:.`/bin/spark-submit --master spark://127.0.0.1:7077 --conf spark.standalone.submit.waitAppCompletion=true --deploy-mode cluster --class org.apache.spark.examples.SparkPi examples/target/original-spark-examples_2.12-3.1.0-SNAPSHOT.jar`

Reply:
Hi @Ngone51 , I tried putting periodic messages in the loop in `pollAndReportStatus` but it doesn't seem to receive message until the loop sending is completed (checked with a `for` loop, will be stuck in an infinite loop in case of current `while(true)` loop). Hence, I have implemented it based on sending an async message from the `pollAndReportStatus` method and if need be, send the message again while receiving the message. Please let me know what you think of this approach. I have tested for the common scenarios and I could see `onNetworkError` method getting called on shutting down Spark master when an application is running.